src/realm/sync/network/default_socket.cpp

```diff
@@ -362,9 +365,153 @@ void DefaultWebSocketImpl::initiate_websocket_handshake()
+        }
+    }
+
+    m_logger_ptr->info("Event loop start()");
+    m_state = Started;
+    auto start_pf = util::make_promise_future<void>();
+    auto stop_pf = util::make_promise_future<void>();
+    m_stop_future = std::move(stop_pf.future);
+    m_thread = std::thread{
+        [this, start_promise = std::move(start_pf.promise), stop_promise = std::move(stop_pf.promise)]() mutable {
+            m_logger_ptr->info("Event loop thread running");
+            // The thread has started - release the main start() future
+            start_promise.emplace_value();
+            auto will_destroy_thread = util::make_scope_exit([&]() noexcept {
+                if (g_binding_callback_thread_observer) {
+                    g_binding_callback_thread_observer->will_destroy_thread();
+                }
+                thread_update_state(Stopped);
+                stop_promise.emplace_value();
+            });
+            if (g_binding_callback_thread_observer) {
+                g_binding_callback_thread_observer->did_create_thread();
+                try {
+                    {
+                        std::lock_guard<std::mutex> lock{m_state_mutex};
+                        // Was the thread stopped before we got here?
+                        if (m_state != Started) {
+                            m_logger_ptr->info("Event loop thread early exit");
+                            return;
+                        }
+                        m_state = Running;
+                    }
+                    m_service.run(); // Throws
+                }
+                catch (const std::exception& e) {
+                    m_logger_ptr->error("Event loop exception: ", e.what());
+                    g_binding_callback_thread_observer->handle_error(e);
+                    throw;
+                }
+            }
+            else {
+                try {
+                    {
+                        std::lock_guard<std::mutex> lock{m_state_mutex};
+                        // Was the thread stopped before we got here?
+                        if (m_state != Started) {
+                            m_logger_ptr->info("Event loop thread early exit");
+                            return;
+                        }
+                        m_state = Running;
+                    }
+                    m_service.run(); // Throws
+                }
+                catch (const std::exception& e) {
+                    m_logger_ptr->error("Event loop exception: ", e.what());
+                    throw;
+                }
+            }
+            m_logger_ptr->info("Event loop thread exiting");
+        }};
+    // Wait for the thread to start before continuing
+    start_pf.future.get();
+    return true;
+}
+
+void DefaultSocketProvider::wait_for_exit()
+{
+    std::unique_lock<std::mutex> lock{m_state_mutex};
+    // do nothing if the thread is not started or the thread stop future is not available
+    if (m_state == NotStarted || !m_stop_future) {
```

Comment:
I think this whole function would be simpler if this was just a state machien where the state eventually == `Stopped`